docs: add howto document for writing feature files

Mohammer5 opened the pull request:
Relates to discussion in dhis2/ui-core#823

varl commented:
Great initiative! :clap: 

This is good to standardize on as it will firstly guide everyone towards using the same language and second, it can be used to resolve discussions. And if it needs to change we will have a paper trail on how this has evolved and why -- which over time is imperative.

I have an ask about the use of the term "SHOULD". Could we use the [RFC keyword definitions](https://tools.ietf.org/html/rfc2119) for "Then" statements?

They have developed excellent definitions for the terms "SHOULD/MUST/MAY" to define specifications, and since our feature files _are_ specs, I think that using those terms and definitions makes sense to avoid ambiguity yet maintain flexibility in writing the acceptance criteria.

Specifically these:

```
1. MUST   This word, or the terms "REQUIRED" or "SHALL", mean that the
   definition is an absolute requirement of the specification.

2. MUST NOT   This phrase, or the phrase "SHALL NOT", mean that the
   definition is an absolute prohibition of the specification.

3. SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.

4. SHOULD NOT   This phrase, or the phrase "NOT RECOMMENDED" mean that
   there may exist valid reasons in particular circumstances when the
   particular behavior is acceptable or even useful, but the full
   implications should be understood and the case carefully weighed
   before implementing any behavior described with this label.

5. MAY   This word, or the adjective "OPTIONAL", mean that an item is
   truly optional.  One vendor may choose to include the item because a
   particular marketplace requires it or because the vendor feels that
   it enhances the product while another vendor may omit the same item.
   An implementation which does not include a particular option MUST be
   prepared to interoperate with another implementation which does
   include the option, though perhaps with reduced functionality. In the
   same vein an implementation which does include a particular option
   MUST be prepared to interoperate with another implementation which
   does not include the option (except, of course, for the feature the
   option provides.)
```

This may be of interest to @vilkg to weigh in on as well so we have a chance to align.

varl commented:
Oh, and second, I think this is a good candidate for a tutorial on developers.dhis2.org.

Mohammer5 commented:
> I have an ask about the use of the term "SHOULD". Could we use the RFC keyword definitions for "Then" statements?

I like! Will include a link and update the howto document

Mohammer5 commented:
> I think this is a good candidate for a tutorial on developers.dhis2.org.

@varl you mean a guide? Should we then remove it from `notes` and move it to developers.dhis2.org?

vilkg commented on howto/writing_feature_files.md:
```diff
@@ -0,0 +1,147 @@
+of what happened to get there. The user could've logged in himself, he could've
+been logged in automatically as his session is still valid, he could've used
+oAuth2 or a simple basic auth. Regardless of how the user has been logged in,
+the state is clear: The user is logged in.
```
I would suggest adding more arguments on why the wrong example is wrong. Some could argue that `Given the user logged in` doesn't express the means of logging in either, just that the user is logged in (state).

Mohammer5 replied:
@vilkg I've added a bit more to make it a bit more clear, did that help or is it still unclear?

vilkg replied:
Yeah, second example is much better @Mohammer5 !

cooper-joe commented:
Thanks for writing up @Mohammer5, this is great to have. My feedback:

- Do we want to define some guidelines for writing `Scenario` titles? @Mohammer5  gave me some good feedback that I always refer to: "Scenario descriptions should be the surrounding context, not the implications". For example:
```
#WRONG
Scenario: Move item from list A to list B

#CORRECT
Scenario: The user clicks the 'move to list B' button
```

- Agreed with @varl about the terminology definitions. I would like to see these words, `MUST, SHOULD, MAY...`, defined clearly in our guidelines. (e.g. it should be clear that `SHOULD != nice to have`, but rather the proper definition of "there may exist valid reasons in particular circumstances to ignore a particular item...")

Mohammer5 commented:
@varl @cooper-joe I've added the modal verbs which have to be used in the `Then` steps, copied the 5 possible verbs and their use-cases and a link to the source

<hr />

@HendrikThePendric I've added definitions for the grammar aspects, including the source.

<hr />

> But in most cases omitting the word user will automatically make this a passive sentence, since In a sentence written in the active voice, the subject of the sentence performs the action.

@HendrikThePendric the `When` step always has to be active, regardless of the term "user"

<hr />

@cooper-joe I've added a "Scenario title" section

Mohammer5 commented:
Thanks @HendrikThePendric, I've changed the text slightly according to your suggestion
